addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,137 @@
+        link_preview_data = self._get_link_preview_from_url(url)
+
+        if link_preview_data:
+            link_preview = self.sudo().create(link_preview_data)
+            return [('insert', link_preview._link_preview_format()[0])]
+        return False
+
+    @api.model
+    def _get_link_preview_from_url(self, url):
+        try:
+            response = requests.head(url, timeout=3)
+        except requests.exceptions.RequestException:
+            return False
+
+        if response.status_code != requests.codes.ok:
+            return False
+
+        image_mimetype = [
+            'image/bmp',
+            'image/gif',
+            'image/jpeg',
+            'image/png',
+            'image/tiff',
+            'image/x-icon',
+        ]
+        # Content-Type header can return a charset, but we just need the
+        # mimetype (eg: image/jpeg;charset=ISO-8859-1)
+        content_type = response.headers['Content-Type'].split(';')
+        if content_type[0] in image_mimetype:
+            return {
+                'description': False,
+                'image_url': url,
+                'mimetype': content_type[0],
+                'title': url,
+                'type': 'image',
+                'url': url,
+            }
+        if 'text/html' in response.headers['Content-Type']:
+            return self._get_link_preview_from_html(url)
+        return False
+
+    def _get_link_preview_from_html(self, url):
+        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0'}
+        response = requests.get(url, timeout=3, stream=True, headers=headers)
+        parser = etree.HTMLParser(encoding=response.encoding)
+        tree = html.fromstring(response.content, parser=parser)
+        title = tree.xpath('//meta[@property="og:title"]/@content')
+        if title:
+            title = title[0]
+            og_type = tree.xpath('//meta[@property="og:type"]/@content')
+            if og_type:
+                og_type = og_type[0]
+            image_url = tree.xpath('//meta[@property="og:image"]/@content')
+            if image_url:
+                image_url = image_url[0]
+            mimetype = tree.xpath('//meta[@property="og:image:type"]/@content')
+            if mimetype:
+                mimetype = mimetype[0]
+            description = tree.xpath('//meta[@property="og:description"]/@content')
+            if description:
+                description = description[0]
+            return {
+                'description': description or False,
+                'image_url': image_url or False,
+                'mimetype': mimetype or False,
+                'title': title or url,
+                'type': og_type or False,
+                'url': url,
+            }
+        return False
+
+    def _delete_and_notify(self):
+        for link_preview in self:
+            self.env['bus.bus']._sendone(self.env.user.partner_id, 'mail.link_preview/delete', {'id': link_preview.id})
```

Comment:
Should use sendmany rather than sendone in loop.

Also this should be sent to all members of the channel when being removed from a channel.